aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -571,95 +571,106 @@ Tensor sparse_compressed_to_dense(
-    auto dst_transposed_sizes = self.sizes().vec();
-    std::swap(dst_transposed_sizes[batch_ndim], dst_transposed_sizes[batch_ndim + 1]);
-    // TODO: write a utility function, or use a transpose once view semantics are there
-    const auto to_transposed_csr = at::native::_sparse_csr_tensor_unsafe(
-        self.ccol_indices(),
-        self.row_indices(),
-        self.values(),
-        dst_transposed_sizes,
-        self.values().scalar_type(),
-        kSparseCsr,
-        self.values().device());
-    auto dst_transposed = at::zeros(dst_transposed_sizes, self.options().layout(kStrided));
-    dst_transposed.add_(to_transposed_csr);
-    return dst_transposed.transpose(batch_ndim, batch_ndim + 1);
-  }
-  if (self.layout() == kSparseBsr || self.layout() == kSparseBsc) {
     Tensor compressed_indices;
     Tensor plain_indices;
     std::tie(compressed_indices, plain_indices) =
         sparse_csr::getCompressedPlainIndices(self);
 
     auto values = self.values();
     Tensor dense = at::zeros(self.sizes(), self.options().layout(kStrided));
-    if (self.dim() == 2) {
-      // Pad shape so we can treat 2-d like batched, we will squeeze out the
-      // phantom batch dim at the end
+
+    if (batch_ndim == 0) {
+      // Pad shape so we can treat non-batched like batched, we will
+      // squeeze out the phantom batch dim at the end.
       compressed_indices.unsqueeze_(0);
       plain_indices.unsqueeze_(0);
-      values = values.unsqueeze_(0);
-      dense = dense.unsqueeze_(0);
+      values.unsqueeze_(0);
+      dense.unsqueeze_(0);
     }
-    if (self.dim() > 3) {
+    if (batch_ndim > 1) {
       // Flatten batch dims
       compressed_indices = compressed_indices.flatten(0, batch_ndim - 1);
       plain_indices = plain_indices.flatten(0, batch_ndim - 1);
       values = values.flatten(0, batch_ndim - 1);
       dense = dense.flatten(0, batch_ndim - 1);
     }
 
-    // At this point everything has 3d shape either the batch dim was inserted,
-    // existed already or was flattened from multiple batch dims
-    std::array<int64_t, 2> blocksize = {values.size(-2), values.size(-1)};
+    // At this point there is only one batch dim, existed already or
+    // was flattened from multiple batch dims.  Now, reshape the
+    // resulting dense matrix so that this single batch dim is joined
+    // with sparse dims into a single dim, so that the remaining dims
+    // are only block dims eventually, and then dense dims.
     auto n_batch = values.size(0);
-    // If we already had batch dim(s) and any of them were zero we can take the
-    // early exit.
-    if (n_batch == 0) {
-      return dense.reshape(self.sizes());
+    int64_t nrows, ncols;
+    auto dense_reshaped_sizes = dense.sizes().vec();
+    if (!block_sparse) {
+      nrows = self.size(batch_ndim);
+      ncols = self.size(batch_ndim + 1);
+      dense_reshaped_sizes.erase(dense_reshaped_sizes.begin());
+      dense_reshaped_sizes.erase(dense_reshaped_sizes.begin());
```

Comment:
Besides batch and sparse dims, there could be dense dims in the picture too.  The point of this reshape is that batch and sparse dims are condensed into a single dim, so that resulting matrix values could be set through single `index_add_()` call further below.

Reply:
OK, thanks for pointing out the possibility of dense dims in `dense_reshaped_sizes`.